README.md
# irony isNow
### A library that expresses rough date or time   
# use
To install ironyIsNow, use yarn
```
$ yarn add @irony0901/is-now
```

# Index
* [Examples](#Examples)   
  * [get started](#getStarted)
  * [unit setting](#unitSetting)
  * [set standard date](#setStandardDate)
  * [create instance](#createInstance)
  * [predict](#predict)

# Examples
## getStarted
``` javascript
import {isNow} from '@irony0901/is-now';

/** 
 * The argument can also be a number type. 
 * Example) isNow.about( Date.now() + 60000 )
**/
const now = Date.now();
const future_afterMoment = isNow.about( new Date( now + 60000 ) );
const future_minute = isNow.about( now + 60001 );
const future_hour = isNow.about( now + 60001 * 60 );
const future_day = isNow.about( now + 60001 * 60 * 24 );
const future_month = isNow.about( now + 60001 * 60 * 24 * 30 );
const future_year = isNow.about( now + 60001 * 60 * 24 * 365 );

const past_justBefore = isNow.about( now - 60000 );
const past_minute = isNow.about( now - 60001 );
const past_hour = isNow.about( now - (60001 * 60) );
const past_day = isNow.about( now - (60001 * 60 * 24) );
const past_month = isNow.about( now - (60001 * 60 * 24 * 30) );
const past_year = isNow.about( now - (60001 * 60 * 24 * 365) );


console.log( future_afterMoment )   // 'after a moment'
console.log( future_minute )        // '1 minute later'
console.log( future_hour )          // '1 hour later'
console.log( future_day )           // '1 day later'
console.log( future_month )         // '1 month later'
console.log( future_year )          // '1 year later'

console.log( past_justBefore )      // 'just before'
console.log( past_minute )          // '1 minute ago'
console.log( past_hour )            // '1 hour ago'
console.log( past_day )             // '1 day ago'
console.log( past_month )           // '1 month ago'
console.log( past_year )            // '1 year ago'
setTimeout(() => {
  /**
   * The current time is initialized (Property named standard in the isNow class) 
   * at the time when the 'isNow' object is loaded.
  **/ 
  console.log('[setTimeout]', isNow.about( new Date( now + 60000 ) ) ) // '[setTimeout] after a moment'
}, 2000)

```  

## unitSetting
``` javascript
import {isNow} from '@irony0901/is-now';

isNow.unit.afterMoment = '조금 후';
isNow.unit.minuteLater = '분 후';
isNow.unit.hourLater = '시간 후';
isNow.unit.dayLater = '일 후';
isNow.unit.weekLater = '주 후';
isNow.unit.monthLater = '월 후';
isNow.unit.yearLater = '년 후';

isNow.unit.justBefore = '방금 전';
isNow.unit.minuteAgo = '분 전';
isNow.unit.hourAgo = '시간 전';
isNow.unit.dayAgo = '일 전';
isNow.unit.weekAgo = '주 전';
isNow.unit.monthAgo = '월 전';
isNow.unit.yearAgo = '년 전';

const now = Date.now();
const future_afterMoment = isNow.about( now + 60000 );
const future_minute = isNow.about( now + 60001 );
const future_hour = isNow.about( now + 60001 * 60 );
const future_day = isNow.about( now + 60001 * 60 * 24 );
const future_month = isNow.about( now + 60001 * 60 * 24 * 30 );
const future_year = isNow.about( now + 60001 * 60 * 24 * 365 );

const past_justBefore = isNow.about( now - 60000 );
const past_minute = isNow.about( now - 60001 );
const past_hour = isNow.about( now - (60001 * 60) );
const past_day = isNow.about( now - (60001 * 60 * 24) );
const past_month = isNow.about( now - (60001 * 60 * 24 * 30) );
const past_year = isNow.about( now - (60001 * 60 * 24 * 365) );

console.log( future_afterMoment )   // '조금 후'
console.log( future_minute )        // '1 분 후'
console.log( future_hour )          // '1 시간 후'
console.log( future_day )           // '1 일 후'
console.log( future_month )         // '1 월 후'
console.log( future_year )          // '1 년 후'

console.log( past_justBefore )      // '방금 전'
console.log( past_minute )          // '1 분 전'
console.log( past_hour )            // '1 시간 전'
console.log( past_day )             // '1 일 전'
console.log( past_month )           // '1 월 전'
console.log( past_year )            // '1 년 전'
```

## setStandardDate
``` javascript
import {isNow} from '@irony0901/is-now';

/**
 * The method named about
 * 
**/
const compareNowResult = isNow.about() // same isNow.about( Date.now() )
console.log('compareNowResult is ', compareNowResult) // 'compareNowResult is after a moment'

isNow.standard = new Date(0); // 1970-01-01T00:00:00.000Z
const compareModifiedResult = isNow.about()
console.log('compareModifiedResult is ', compareModifiedResult) // 'compareModifiedResult is ${N} year later'

```

## createInstance
``` javascript
import { isNow, createIsNow } from '@irony0901/is-now';

const isNow_ko = createIsNow();
isNow_ko.unit.afterMoment = '조금 후';
isNow_ko.unit.minuteLater = '분 후';
isNow_ko.unit.hourLater = '시간 후';
isNow_ko.unit.dayLater = '일 후';
isNow_ko.unit.weekLater = '주 후';
isNow_ko.unit.monthLater = '월 후';
isNow_ko.unit.yearLater = '년 후';

isNow_ko.unit.justBefore = '방금 전';
isNow_ko.unit.minuteAgo = '분 전';
isNow_ko.unit.hourAgo = '시간 전';
isNow_ko.unit.dayAgo = '일 전';
isNow_ko.unit.weekAgo = '주 전';
isNow_ko.unit.monthAgo = '월 전';
isNow_ko.unit.yearAgo = '년 전';

console.log('Ver_Eng', isNow.about())   // 'Ver_Eng after a moment'
console.log('Ver_Ko', isNow_ko.about()) // 'Ver_Ko 조금 후'

```

## predict
A method that can modify the return result of the `'about'` method.   
The initialization form is as follows
```
({diff, about, unit}: ElapsedPredictArg) => 
  Math.abs(diff) === about ? unit : `${about} ${unit}`
```
### ElapsedPredictArg
* diff: number   
  The difference between the time at the time of the `'about'` method and the time of the [`IsNow.standard`](#setStandardDate) field.   
  Millisecond.   
  positive value - future   
  negative value - past
* about: number   
  Unconditionally positive value.   
  Value to be expressed according to the value of the field in [`IsNow.unit`](#unitSetting)
* unit: string   
  The value of one of the fields in [`IsNow.unit`](#unitSetting)   

``` javascript
import {isNow} from '@irony0901/is-now';

const now = Date.now();
isNow.predict = ({diff, about, unit}) => {
  console.log(`[PREDICT] diff: ${diff}, about: ${about}, unit: ${unit}`);
  return Math.abs(diff) === about ? unit : `${about} ${unit}`
}

console.log('[afterMoment]', isNow.about());
/** 
 * '[PREDICT] diff: 1, about: 1, unit: after a moment'
 * '[afterMoment] after a moment'
 **/

console.log('[minute]', isNow.about(  now + (12000 * 2) ) );
/** 
 * '[PREDICT] diff: 120007, about: 2, unit: minute later'
 * '[minute] 2 minute later'
 **/

isNow.predict = ({diff, about, unit}) => {
  return Math.abs(diff) === about ? `${unit}. [${about}] millisconds` : `[${about}] ${unit}`
}

console.log('[afterMoment]', isNow.about(now + 250));
/** 
 * '[afterMoment] after a moment. [250] milliscdonds'
 **/

console.log('[minute]', isNow.about(  now + (12000 * 2) ) );
/** 
 * '[minute] [2] minute later'
 **/
```

CHANGELOG.md
# IsNow change log

## 1.0.3
* remove import default.

## 1.0.2
* Modify the wrong word from README. 'Isnow.strandard' ->  'IsNow.standard'
* The build result in the browser is umd.

## 1.0.1
* Modify the wrong word from README.

## 0.3.2
* Script command change.
* Remove the @types/uuid from devDependencies
* modify value. ElapsedPredictArg.diff.  
  The value of the amount is the future. The value of the negative is the past
* predict example addition.
* Remove devDependency of rollup-typescript


rollup.config.js
import pkg from "./package.json";
import pluginTypescript from "rollup-plugin-typescript2";
import pluginCommonjs from "@rollup/plugin-commonjs";
import pluginNodeResolve from "@rollup/plugin-node-resolve";
import { babel } from "@rollup/plugin-babel";
import * as path from "path";
const moduleName = pkg.name.replace(/^@.*\//, '');
const inputFileName = "src/index.ts";
const author = pkg.author;
const banner = `
  /**
   * @license
   * author: ${author.name}
   * ${moduleName}.js v${pkg.version}
   * Released under the ${pkg.license} license.
   */
`;
const typescriptConfig = {
  useTsconfigDeclarationDir: false,
  // declarationDir: './types',
  exclude: [ 'test/**/*.ts' ]
}
export default [

  {
    input: inputFileName,
    output: [
      {
        name: moduleName.replace('-', '_'),
        file: pkg.browser,
        format: "umd",
        // sourcemap: "inline",
        // exports: "named",
        banner,
      }
      // {
      //   name: moduleName.replace('-', '_'),
      //   file: pkg.browser.replace(".js", ".min.js"),
      //   format: "iife",
      //   // sourcemap: "inline",
      //   exports: "named",
      //   banner,
      //   // plugins: [terser()],
      // },
    ],
    // preserveModules: true,
    plugins: [
      pluginTypescript(typescriptConfig),
      pluginCommonjs({
        extensions: [".js", ".ts"],
      }),
      babel({
        babelHelpers: "bundled",
        configFile: path.resolve(__dirname, ".babelrc.js"),
      }),
      pluginNodeResolve({
        browser: true,
      }),
    ],
  },

  // ES
  {
    input: inputFileName,
    output: [
      {
        // dir: 'dist',
        file: pkg.module,
        format: "es",
        // sourcemap: "inline",
        // sourcemap: true,
        banner,
        exports: "named",
      },
    ],
    external: [
      ...Object.keys(pkg.dependencies || {}),
      ...Object.keys(pkg.devDependencies || {}),
    ],
    // preserveModules: true,
    plugins: [
      pluginTypescript(typescriptConfig),
      pluginCommonjs({
        extensions: [".js", ".ts"],
      }),
      babel({
        babelHelpers: "bundled",
        configFile: path.resolve(__dirname, ".babelrc.js"),
      }),
      pluginNodeResolve({
        browser: false,
      }),
    ],
  },

  // // CommonJS
  {
    input: inputFileName,
    output: [
      {
        file: pkg.main,
        format: "cjs",
        // sourcemap: "inline",
        banner,
        exports: "default",
      },
    ],
    external: [
      ...Object.keys(pkg.dependencies || {}),
      ...Object.keys(pkg.devDependencies || {}),
    ],
    plugins: [
      pluginTypescript(typescriptConfig),
      pluginCommonjs({
        extensions: [".js", ".ts"],
      }),
      babel({
        babelHelpers: "bundled",
        configFile: path.resolve(__dirname, ".babelrc.js"),
      }),
      pluginNodeResolve({
        browser: false,
      }),
    ],
  },
];

src/is_now.ts
export const ONE_MINUTE = 60 * 1000,
  ONE_HOUR = ONE_MINUTE * 60,
  ONE_DAY = ONE_HOUR * 24,
  ONE_WEEK = ONE_DAY * 7,
  ONE_MONTH = ONE_DAY * 30,
  ONE_YEAR = ONE_DAY * 365;

type ElapsedUnit = {
  yearAgo: string;
  monthAgo: string;
  weekAgo: string;
  dayAgo: string;
  hourAgo: string;
  minuteAgo: string;
  justBefore: string;

  yearLater: string;
  monthLater: string;
  weekLater: string;
  dayLater: string;
  hourLater: string;
  minuteLater: string;
  afterMoment: string;
}
export type ElapsedPredictArg = {
  diff: number;
  about: number;
  unit: string;
}
export type ElapsedPredict = (arg: ElapsedPredictArg) => string;

class IsNow {
  private standard_: Date;
  private standardLong: number;
  public unit: ElapsedUnit = {
    yearAgo: 'year ago',
    monthAgo: 'month ago',
    weekAgo: 'week ago',
    dayAgo: 'day ago',
    hourAgo: 'hour ago',
    minuteAgo: 'minute ago',
    justBefore: 'just before',

    yearLater: 'year later',
    monthLater: 'month later',
    weekLater: 'week later',
    dayLater: 'day later',
    hourLater: 'hour later',
    minuteLater: 'minute later',
    afterMoment: 'after a moment'
  };
  private predict_: ElapsedPredict = ({diff, about, unit}) => 
    Math.abs(diff) === about ? unit : `${about} ${unit}`
  ;

  constructor( standard: Date = new Date ){
    this.standard = standard;
    // this.standardLong = standard.getTime();
  }

  
  set predict( predict: ElapsedPredict ){
    this.predict_ = predict;
  }
  get predict(){
    return this.predict_;
  }

  set standard(standard: Date) {
    this.standard_ = standard
    this.standardLong = standard.getTime();
  }
  get standard() {
    return this.standard_;
  }

  about(dateOrMilliseconds: Date|number = new Date()): string {
    if( !((Number(dateOrMilliseconds) === dateOrMilliseconds) || dateOrMilliseconds instanceof Date)  )
      return undefined;

    const compare: number = 
      dateOrMilliseconds instanceof Date 
      ? dateOrMilliseconds.getTime()
      : dateOrMilliseconds
    ;
    const diff = compare - this.standardLong;
    if( diff < 1 ){

      const positiveDiff = Math.abs(diff);
      if(positiveDiff > ONE_YEAR)
        return this.predict({diff, about: Math.floor(positiveDiff / ONE_YEAR), unit: this.unit.yearAgo})
        // return `${Math.floor(diff / ONE_YEAR)} ${this.unit.yearAgo}`
      else if(positiveDiff > ONE_MONTH)
        return this.predict({diff, about: Math.floor(positiveDiff / ONE_MONTH), unit: this.unit.monthAgo})
        // return `${Math.floor(diff / ONE_MONTH)} ${this.unit.monthAgo}`
      else if(positiveDiff > ONE_WEEK)
        return this.predict({diff, about: Math.floor(positiveDiff / ONE_WEEK), unit: this.unit.weekAgo})
      // return `${Math.floor(diff / ONE_WEEK)} ${this.unit.weekAgo}`
      else if(positiveDiff > ONE_DAY)
        return this.predict({diff, about: Math.floor(positiveDiff / ONE_DAY), unit: this.unit.dayAgo})
        // return `${Math.floor(diff / ONE_DAY)} ${this.unit.dayAgo}`
      else if(positiveDiff > ONE_HOUR)
        return this.predict({diff, about: Math.floor(positiveDiff / ONE_HOUR), unit: this.unit.hourAgo})
        // return `${Math.floor(diff / ONE_HOUR)} ${this.unit.hourAgo}`
      else if(positiveDiff > ONE_MINUTE)
        return this.predict({diff, about: Math.floor(positiveDiff / ONE_MINUTE), unit: this.unit.minuteAgo})
        // return `${Math.floor(diff / ONE_MINUTE)} ${this.unit.minuteAgo}`
      else 
        return this.predict({diff, about: positiveDiff, unit: this.unit.justBefore})
        // return this.unit.justBefore
      
    }else{

      // diff = Math.abs(diff);
      
      if(diff > ONE_YEAR)
        return this.predict({diff, about: Math.floor(diff / ONE_YEAR), unit: this.unit.yearLater})
        // return `${Math.floor(diff / ONE_YEAR)} ${this.unit.yearLater}`
      else if(diff > ONE_MONTH)
        return this.predict({diff, about: Math.floor(diff / ONE_MONTH), unit: this.unit.monthLater})
        // return `${Math.floor(diff / ONE_MONTH)} ${this.unit.monthLater}`
      else if(diff > ONE_WEEK)
        return this.predict({diff, about: Math.floor(diff / ONE_WEEK), unit: this.unit.weekLater})
        // return `${Math.floor(diff / ONE_WEEK)} ${this.unit.weekLater}`
      else if(diff > ONE_DAY)
        return this.predict({diff, about: Math.floor(diff / ONE_DAY), unit: this.unit.dayLater})
        // return `${Math.floor(diff / ONE_DAY)} ${this.unit.dayLater}`
      else if(diff > ONE_HOUR)
        return this.predict({diff, about: Math.floor(diff / ONE_HOUR), unit: this.unit.hourLater})
        // return `${Math.floor(diff / ONE_HOUR)} ${this.unit.hourLater}`
      else if(diff > ONE_MINUTE)
        return this.predict({diff, about: Math.floor(diff / ONE_MINUTE), unit: this.unit.minuteLater})
        // return `${Math.floor(diff / ONE_MINUTE)} ${this.unit.minuteLater}`
      else 
        return this.predict({diff, about: diff, unit: this.unit.afterMoment})
        // return this.unit.afterMoment
      
    }
  }
}


export const isNow = new IsNow();

export const createIsNow = ( standard: Date = new Date() ) => new IsNow( standard );